### ISSUE TITLE
Discarding frames with too much motion blur completely

### ISSUE THREAD

**Issue by lvml** [georgmartius/vid.stab#24]
Hi Georg,

first of all: Thanks a lot for libvidstab, which I am using to my greatest pleasure via ffmpeg! I had tried it long ago with some 1080p material, with mixed results, but recently found that when I use vidstab on 4k videos I recorded, the results are really great, probably because there's so much "spare resolution" in the recordings that a little zooming doesn't hurt the looks of the result as much as with 1080p material.

But now back on topic: I use libvidstab for one additional purpose you probably hadn't in mind when you wrote it: For creating low-noise still images from hand-held video shots under poor lighting conditions. The idea here is that if you don't have a tripod, but want to take a picture in low light, you can substitute for the noise-cancellation that is usually done in-camera for long-exposure still images by first recording some seconds of hand-held 4k-video with relatively high ISO values, then software-stabilize the clip using libvidstab (in "tripod=1" mode), then use a filter like ffmpeg's "hqdn3d" with spatial=0 but temporal=200 or alike parametrization, effectively doing the same "denoising" the camera could have done when using a tripod but from the stabilized hand held video, and save a resulting frame at the end of that clip as a still image.

This already has worked for me remarkebly well, however, there's one catch I'd like to get rid of: If inside the hand-held video there are individual frames with too much motion-blur from unintentional movements, they can significantly lower the quality of the overall result.

Now I wonder whether this could be easily solved by just discarding all frames from the video that have too much motion blur in them.

I understand that libvidstab already does some contrast analysis for the "mincontrast" parameter, and maybe that could be a starting point to implement a minimum threshold for micro-contrast that, if undercut, could cause libvidstab to ignore the whole frame not only for motion estimation, but to drop the frame completely (or replace it by a duplicate of the preceding frame, if that is easier to implement).

What do you think, is such a feature feasible with reasonably low effort?

**Comment by magicgoose** [comment]
ffmpeg then would have to deal with variable frame rate…
or, for this specific scenario, after skipped frame all subsequent frames can be shifted back in time. I don't know if this is easier, though

**Comment by georgmartius** [comment]
On Monday 23 March 2015 23:41:30 magicgoose wrote:
> ffmpeg then would have to deal with variable frame rate…
I think the only reasonable solution would be to double the frame and move it 
with the average camera velocity. 

> or, for this specific scenario, after skipped frame all subsequent frames
> can be shifted back in time. I don't know if this is easier, though
We would get sync problems with audio etc.

**Comment by magicgoose** [comment]
> We would get sync problems with audio etc.

Of course, but _in OP's case_ this is going to be useful and obviously he doesn't care about time at all — the final result is still picture

**Comment by lvml** [comment]
Indeed, neither movement nor audio would be of concern, and duplicating a preceding frame when discarding one would be good enough for the purpose.

That's why I think this feature could be realized with reasonably low effort, and open up a great opportunity to do spontaneous night-shots without tripods, even with not-so-big-sensor cameras. :-)

**Comment by magicgoose** [comment]
duplicating is worse for OP, because then it (the doubled frame) will have bigger "weight" in averaging, making noise reduction less efficient

**Comment by lvml** [comment]
On 03/25/2015 12:34 PM, magicgoose wrote:
> duplicating is worse for OP, because then it (the doubled frame) will have bigger "weight" in
> averaging, making noise reduction less efficient

Yes, if it was possible to drop a frame completely (essentially making
the output video "shorter") that would be even better.

**Comment by georgmartius** [comment]
On Wednesday 25 March 2015 04:39:23 Lutz Vieweg wrote:
> On 03/25/2015 12:34 PM, magicgoose wrote:
> > duplicating is worse for OP, because then it (the doubled frame) will have
> > bigger "weight" in averaging, making noise reduction less efficient
> 
> Yes, if it was possible to drop a frame completely (essentially making
> the output video "shorter") that would be even better.
I understand. I have to check whether I can do it in ffmpeg easily. 
On the other hand, I don't like to spend too much time with a very specific 
feature for one type of shoots. 

Can you open a feature request on github for it.

Georg
> 
> 
> 
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/georgmartius/vid.stab/issues/24#issuecomment-85992679

-- 
---- Georg Martius,  Tel: +49 177 6413311  -----
------- http://georg.hronopik.de ----------

**Comment by lvml** [comment]
> Can you open a feature request on github for it.

I thought this (#24) was the recommended way to submit a feature request... or what other method is preferred?

**Comment by georgmartius** [comment]
Sorry, my mistake. This is perfectly fine.

**Comment by lvml** [comment]
Here is a sample to demonstrate how well the approach already works:

This picture was taken under very difficult lighting conditions, free hand, using ISO 400. It exposes both some image noise (which was reduced, though, when generating the JPEG from the RAW) and at the same time, it was not possible to avoid motion-blur when shooting free-handed (enlarge to watch 1:1 zoom): 

<img src="http://filebin.net/eoncme5t68/denoise_pic_iso_400.jpg" width=640>

The following picture in contrast is a screenshot (at position 9 seconds from start) taken from <a href="linktobeinsertedafteruploadfinished">this original 3840x2160 video</a>, which was taken under the same difficult lighting conditions at ISO 6400, after stabilizing and temporal denoising using <a href="http://filebin.net/eoncme5t68/stabilize_static_denoise">this script</a> (enlarge to watch 1:1 zoom):

<img src="http://filebin.net/eoncme5t68/denoise_pic_from_vidstab.jpg">

As you can see, the screen shot from the video is the much better still image.

But the sample video contains a (deliberate) strong movement at 10s into the video, and if you watch the resulting video after stabilization / temporal denoising, you can see how the quality suffers from that. And in many cases you cannot avoid having frequent shakes when shooting free hand. So it would be very good if vidstab could just discard the few frames in the middle of the sample .MP4 file where there is such strong movement / motion blur.

**Comment by georgmartius** [comment]
Closing: **not planned.**

Dropping frames is outside what this library does. vid.stab produces one transform per input frame and hands the frames back in the same order and number — it never changes the frame sequence. Making it drop frames would break that contract in a way the two-pass design cannot absorb: the transform file is positional, one entry per frame consumed in order, so a second pass that emitted fewer frames than the first analysed would put everything downstream out of step (the failure mode in #127). Frame-rate and frame-count decisions belong in the filter chain around us, not inside us.

ffmpeg already has the pieces, and they compose properly because they run in *both* passes:

```shell
# drop near-duplicate frames
ffmpeg -i in.mp4 -vf mpdecimate,vidstabdetect=result=t.trf -f null -
ffmpeg -i in.mp4 -vf mpdecimate,vidstabtransform=input=t.trf out.mp4
```

For blur specifically there is no ready-made ffmpeg filter, but a blur metric is straightforward to compute externally (variance of the Laplacian is the usual one) and turn into a `select` expression or a frame list — and doing it as a separate step lets you tune the threshold against your footage, which is the part that actually decides whether the result looks better or just choppier.

That last point is the substantive objection, beyond the architectural one: replacing a blurred frame with a held or interpolated one is often more noticeable than the blur was, especially when blur comes in runs rather than isolated frames. `minterpolate` is the better tool if you want to synthesize replacements.

Thanks for the long discussion in this thread — closing so the open list reflects what is actually intended.
